ORF discovery tools
==========

ORF-RATER
----------

Input:
Reference transcriptomes (NOT whole genome), previously assembled + RefSeq + ENSEMBL transcripts

Output:

/////////////////////////
Dependency:

* plastid package
Performing metagene analyses - https://plastid.readthedocs.org/en/latest/examples/metagene.html
Determine P-site offsets for ribosome profiling data - https://plastid.readthedocs.org/en/latest/examples/p_site.html#determine-p-site-offsets-for-ribosome-profiling-data
Read phasing in ribosome profiling - https://plastid.readthedocs.org/en/latest/examples/phasing.html

/////////////////////////
Scripts:

* prune_transcripts.py

Use ribosome profiling data to remove unwanted transcripts from a
transcriptome. Transcripts will be removed if they do not have enough reads of
the desired length, if they have greater than some fraction of their reads
from a single position, if they are annotated as pseudogenes (and multimap
with another transcript), or if they have more multimapping reads beyond what
would be expected based on the number of multimapping positions. It is
recommended that this file be run in an empty directory and that OUTBED remain
at the default value ("transcripts.bed") for consistency with later scripts.

transcriptome-aligned BAM file(s), Ideally, should be ribosome profiling data 
sets collected in the absence of initiation inhibitors (e.g. CHX or no drug)

fpsize = opts.minlen
psite = (fpsize + 1) / 2

* make_tfams.py

Identify overlapping transcripts from a bed file, termed transcript families
("tfams"). Each tfam will be assigned a name, either based on the transcript
IDs or from an optional gene name file. It is recommended that this script be
run in a newly created folder containing only the input BED file, generated by
prune_transcripts.py, and possibly the GENENAMES file. It is also recommended
that the default value ("tfams") be used for parameter TFAMSTEM, for
consistency with later scripts.

* find_orfs_and_types.py

Identify all possible ORFs in a transcriptome. ORF-RATER will evaluate translation 
of only these ORFs. Default start codon "ATG", can be "NTG"


/////////////////////////
Workflow Steps:
1. 

